### PR TITLE
functional-tester: add lease stresser

### LINKED ISSUE #6380: functional-tests: stress leases
We would like to ensure lease related operations are handled correctly in failure recovery path but currently there is no lease testing in the functional tester.

Invariants to test:
* Lease should not expire before the returned TTL
* Any key attached to an expired/revoked lease should be removed
* All members key + lease state should have same hash for the same revision

Implementation would be a new stresser, `leaseStresser`:
* creates N lease and attach M keys onto each lease it creates. Sends keeps alive to all created leases.
* randomly the leaseStresser drops leases from the set of leases it keeps; creates new lease to replace it.
* checks the dropped leases are expired, and the keys attached to them are deleted.
* checks the invariants

Work break down

1. Setup etcd locally with 3 etcd clusters including agent
2. Run functional tester on the local etcd clusters
3. Modify the code for lease testing
4. Testing the modification locally.
5. Send PR for review

### PR DESCRIPTION
**Pull request by fanminshi**
Add lease stresser to test lease code under stress and etcd failures

resolve #6380

### CODE REVIEW

**Comment by fanminshi** [comment]
base on https://github.com/coreos/etcd/pull/6588

**Comment by fanminshi** [comment]
PTAL cc @xiang90 @gyuho

File: tools/functional-tester/etcd-tester/checks.go [hunk 1]
```diff
@@ -69,6 +74,100 @@ func (hc *hashChecker) Check() (err error) {
...
+	plog.Info("lease stresser invariant check...")
+	errc := make(chan error)
+	for _, ls := range lc.leaseStressers {
+		go func(s Stresser) { errc <- lc.checkInvariant(s) }(ls)
+	}
+	var errs []error
+	for i := 0; i < len(lc.leaseStressers); i++ {
+		if err := <-errc; err != nil {
+			errs = append(errs, err)
+		}
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/checks.go]:
```go
if i == len(lc.leaseStressers) - 1 {
    close(errc)
}
```

to be more Go-idiomatic on ranging channels?

File: tools/functional-tester/etcd-tester/checks.go [hunk 2]
```diff
@@ -34,6 +37,8 @@ type hashChecker struct {
 
 func newHashChecker(hrg hashAndRevGetter) Checker { return &hashChecker{hrg} }
 
+const leaseCheckerIimeout = 10 * time.Second
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: Iimeout -? Timeout

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
+// Copyright 2015 The etcd Authors
```
Comment (gyuho) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: use `2016`

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// time to live for lease
+const TTL = 5
```
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: Anthony recommends 30. In that case, it is very unlikely for lease timeout by its own

File: tools/functional-tester/etcd-tester/checks.go (hunk 2 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: good catch!

File: tools/functional-tester/etcd-tester/stresser.go [hunk 5]
```diff
@@ -165,16 +165,11 @@ func (s *stresser) Stress() error {
 	if err != nil {
 		return fmt.Errorf("%v (%s)", err, s.Endpoint)
 	}
-	ctx, cancel := context.WithCancel(context.Background())
 
 	wg := &sync.WaitGroup{}
 	wg.Add(s.N)
-
-	s.mu.Lock()
```
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/stresser.go]: apparently, stress() will be called serially not concurrently. so no need to use lock()

Comment (fanminshi) [inline tools/functional-tester/etcd-tester/stresser.go]: will do

File: tools/functional-tester/etcd-tester/lease_stresser.go (hunk 4 above)
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: sure.

**Comment by xiang90** [comment]
LGTM

File: tools/functional-tester/etcd-tester/checks.go (hunk 1 above)
Comment (gyuho) [inline tools/functional-tester/etcd-tester/checks.go]: I thought this was receiving from channel in for-loop. nvm

**Comment by fanminshi** [comment]
ALL fixed

**Pull request merged**